src/urbit.ts

```diff
@@ -0,0 +1,468 @@
+          //         path: '/~zod/mailbox'
+          //       }
+          //     }
+          //   }
+          // }
+        }
+      });
+      this.sseClient?.addEventListener('error', function(event: Event) {
+        console.error('pipe error', event);
+      });
+    }
+    return this.sseClient;
+  }
+
+  /**
+   * Autoincrements the next event ID for the appropriate channel.
+   */
+  getEventId(): number {
+    this.lastEventId = Number(this.lastEventId) + 1;
+    return this.lastEventId;
+  }
+
+  /**
+   * Acknowledges an event.
+   *
+   * @param eventId The event to acknowledge.
+   */
+  async ack(eventId: number): Promise<void | number> {
+    return this.sendMessage('ack', { 'event-id': eventId });
+  }
+
+  /**
+   * This is a wrapper method that can be used to send any action with data.
+   *
+   * Every message sent has some common parameters, like method, headers, and data
+   * structure, so this method exists to prevent duplication.
+   *
+   * @param action The action to send
+   * @param data The data to send with the action
+   * 
+   * @returns void | number If successful, returns the number of the message that was sent
+   */
+  async sendMessage(action: string, data?: object): Promise<void | number> {
+    const headers: headers = {
+      'Content-Type': 'application/json',
+    };
+    if (!isBrowser) {
+      headers.Cookie = this.cookie;
+    }
+    const id = this.getEventId();
+    if (this.verbose) {
+      console.log(`Sending message ${id}:`, action, data, headers);
+    }
+    return fetch(this.channelUrl, {
+      method: 'put',
+      headers,
+      credentials: 'include',
+      body: JSON.stringify([{
+        id,
+        action,
+        ...data,
+      }])
+    }).catch((error: any) => {
+      console.error('message error', error);
+    })
+    .then(response => {
+      if (this.verbose) {
+        console.log(`Received from message ${id}: `, response);
+      }
+      return id;
+    });
+  }
+
+  /**
+   * Pokes a ship with data.
+   *
+   * @param ship The ship to poke
+   * @param app The app to poke
+   * @param mark The mark of the data being sent
+   * @param json The data to send
+   */
+  async poke(
+    ship: string = this.ship ? this.ship : '',
+    app: string,
+    mark: string,
+    json: Object,
+    successFunc = () => {},
+    failureFunc = () => {}
+  ): Promise<void | number> {
+    const pokeId = await this.sendMessage('poke', { ship, app, mark, json });
+
+    if (!pokeId) return;
+
+    this.outstandingPokes.set(pokeId, {
+      success: successFunc,
+      fail: failureFunc
+    });
+
+    return pokeId;
+  }
+
+  /**
+   * Subscribes to a path on an app on a ship.
+   *
+   * @param ship The ship to subscribe to
+   * @param app The app to subsribe to
+   * @param path The path to which to subscribe
+   */
+  async subscribe({
+    ship = this.ship ? this.ship : '',
+    app,
+    path,
+    err = () => {},
+    event = () => {},
+    quit = () => {}
```

Comment:
I'd have SubscriptionInterface as the last argument here, a few too many arguments in this function. Same quibbles around acking and Promises as the poke function too

Reply:
I didn't think you could set defaults with that pattern. Can you give me an example of what you're thinking?

Reply:
```typescript
async subscribe(app: string, path: string, somehandlers: Partial<SubscriptionInterface>) {
   const handlers = { err: () => {}, event: () => {}, quit: () => {}, ...somehandlers };
   /// etc.
}